### PR TITLE
New context action recording

### PR DESCRIPTION
**Pull request by tizoc**
OCaml-side: https://gitlab.com/simplestaking/tezos/-/merge_requests/16

Changes:

- Simplified FFI callbacks. A single "send" function is used now, with "record" and "perform" flags.
- `ContextActionMessage` wrapper for `ContextAction` that carries along these flags, which are checked by the context listener to see what to do. Read actions will contain both flags. Write actions will contain the "record" flag when initially sent  from the OCaml side, and the "perform" flag when replayed (but not the "record" one).
- New `tree_hash` (hash of the tree to which the action is being applied) and `new_tree_hash` (on actions that produce a new tree, this is the hash of the new tree) fields on `ContextAction`.

### CODE REVIEW

**Comment by tizoc** [comment]
@bkontur I guess that because of the (context actions) message format I need to update some other code (or docker image) for drone? It never finished the run and there are a bunch of failures (related to `listen_protocol_events`).

**Comment by tizoc** [comment]
Here http://ci.tezedge.com/simplestaking/tezedge/2312/1/7

```
Jan 23 16:52:48.512 WARN Failed to receive event from protocol runner, reason: DeserializationError { reason: "Io(Kind(UnexpectedEof))" }
91 | Jan 23 16:52:48.514 INFO Context listener finished
92 | Jan 23 16:52:48.515 INFO Waiting for connection from protocol runner
93 | Jan 23 16:52:48.517 WARN Error while processing protocol events, reason: SendError { reason: Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" } }, endpoint: test_actors_apply_blocks_and_check_context_and_mempool_write_runner
94 | Jan 23 16:52:51.532 CRIT Error process context event, reason: AcceptTimeout { timeout: 3s }
95 |  
96 | 0: failure::backtrace::internal::InternalBacktrace::new
97 | 1: failure::backtrace::Backtrace::new
98 | 2: shell::context_listener::listen_protocol_events
99 | 3: std::sys_common::backtrace::__rust_begin_short_backtrace
100 | 4: core::ops::function::FnOnce::call_once{{vtable.shim}}
101 | 5: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
102 | at /rustc/e2267046859c9ceb932abc983561d53a117089f6/library/alloc/src/boxed.rs:1328:9
103 | <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
104 | at /rustc/e2267046859c9ceb932abc983561d53a117089f6/library/alloc/src/boxed.rs:1328:9
105 | std::sys::unix::thread::Thread::new::thread_start
106 | at /rustc/e2267046859c9ceb932abc983561d53a117089f6/library/std/src/sys/unix/thread.rs:71:17
107 | 6: start_thread
108 | 7: clone
109
```

**Comment by bkontur** [comment]
> 
> 
> Here http://ci.tezedge.com/simplestaking/tezedge/2312/1/7
> 
> ```
> Jan 23 16:52:48.512 WARN Failed to receive event from protocol runner, reason: DeserializationError { reason: "Io(Kind(UnexpectedEof))" }
> 91 | Jan 23 16:52:48.514 INFO Context listener finished
> 92 | Jan 23 16:52:48.515 INFO Waiting for connection from protocol runner
> 93 | Jan 23 16:52:48.517 WARN Error while processing protocol events, reason: SendError { reason: Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" } }, endpoint: test_actors_apply_blocks_and_check_context_and_mempool_write_runner
> 94 | Jan 23 16:52:51.532 CRIT Error process context event, reason: AcceptTimeout { timeout: 3s }
> 95 |  
> 96 | 0: failure::backtrace::internal::InternalBacktrace::new
> 97 | 1: failure::backtrace::Backtrace::new
> 98 | 2: shell::context_listener::listen_protocol_events
> 99 | 3: std::sys_common::backtrace::__rust_begin_short_backtrace
> 100 | 4: core::ops::function::FnOnce::call_once{{vtable.shim}}
> 101 | 5: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
> 102 | at /rustc/e2267046859c9ceb932abc983561d53a117089f6/library/alloc/src/boxed.rs:1328:9
> 103 | <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
> 104 | at /rustc/e2267046859c9ceb932abc983561d53a117089f6/library/alloc/src/boxed.rs:1328:9
> 105 | std::sys::unix::thread::Thread::new::thread_start
> 106 | at /rustc/e2267046859c9ceb932abc983561d53a117089f6/library/std/src/sys/unix/thread.rs:71:17
> 107 | 6: start_thread
> 108 | 7: clone
> 109
> ```

hmm , I saw quickly the PR, and looks ok,
hmm there is a error at the begining - the first `DeserializationError`:
```
Jan 23 16:52:38.173 INFO Waiting for connection from protocol runner
Jan 23 16:52:38.173 INFO Received connection from protocol runner. Starting to process context events.
Jan 23 16:52:38.225 DEBG Connection for protocol runner was created successfully, endpoint: test_actors_apply_blocks_and_check_context_and_mempool_readonly_runner_pool_1
Jan 23 16:52:38.325 WARN Failed to receive event from protocol runner, reason: DeserializationError { reason: "Io(Kind(UnexpectedEof))" }
Jan 23 16:52:38.325 INFO Context listener finished
Jan 23 16:52:38.325 INFO Waiting for connection from protocol runner
Jan 23 16:52:38.326 WARN Error while processing protocol events, reason: SendError { reason: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" } }, endpoint: test_actors_apply_blocks_and_check_context_and_mempool_write_runner
```
well, something is wrong, 

did you try to run it locally?
```
cd tezedge
cargo clean
cargo build --release
PROTOCOL_RUNNER=../target/release/protocol-runner cargo test --release -- --nocapture --ignored test_actors_apply_blocks_and_check_context_and_mempool
```

**Comment by tizoc** [comment]
Ok, I will do the clean + compile in release mode now and will run the test once I get back.

**Comment by tizoc** [comment]
Just did a quick test and I was able to reproduce it locally with the commands you pasted. Will try to debug it tomorrow.

File: shell/src/context_listener.rs [hunk 1]
```diff
@@ -317,3 +250,83 @@ fn listen_protocol_events(
...
+                HashType::ContextHash.hash_to_b58check(new_context_hash),
+                HashType::ContextHash.hash_to_b58check(&hash),
+            );
+        }
+
+        ContextAction::Checkout { context_hash, .. } => {
+            context.checkout(context_hash)?;
+        }
+
+        ContextAction::Commit { .. } => (), // Ignored (no block_hash)
```
Comment (tizoc) [inline shell/src/context_listener.rs:324]: @bkontur on which situations can this happen? (this was implicit in the old version, but I removed the wildcard match, which I think we should avoid as much as possible).

File: shell/src/context_listener.rs [hunk 2]
```diff
@@ -317,3 +250,83 @@ fn listen_protocol_events(
...
+
+        ContextAction::Fold { .. } => (), // Ignored
```
Comment (tizoc) [inline shell/src/context_listener.rs:326]: This one is ignored because FOLDs are not supported yet.

File: shell/src/context_listener.rs (hunk 1 above)
Comment (bkontur) [reply shell/src/context_listener.rs:324]: hmm, I dont know, it should not happen, even commit_genesis should be with block_hash:
```
Option.iter
    (fun inspector ->
      inspector#mark_header_and_protocol chain_genesis.block chain_genesis.protocol)
    context_raw_inspector ;
```

**Comment by tizoc** [comment]
@bkontur found the problem (wish typecheckers worked across processes!), lets see if it runs now: https://github.com/simplestaking/tezedge/pull/376/commits/d3847e19a7ecf6b739b52b417f7b3bc713de21b7

**Comment by tizoc** [comment]
It failed the python tests, will see if I can figure out why.

**Comment by tizoc** [comment]
Closing this one, replaced by #383

**Pull request closed without merging**